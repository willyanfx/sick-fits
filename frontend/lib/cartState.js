import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();

const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = () => {
    setCartOpen(!cartOpen);
  };

  const closeCart = () => {
    setCartOpen(false);
  };

  const openCart = () => {
    setCartOpen(true);
  };
  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
