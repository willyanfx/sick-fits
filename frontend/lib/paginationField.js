import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      const { skip, first } = args;

      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing.slice(skip, skip + first).filter((x) => x);
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        // no items, fetch them
        return false;
      }

      if (items.length) {
        console.log(
          `there are ${items.length} items in the cache! gonna send them to apollo`
        );
        return items;
      }

      return false;

      // ask the read -> return the items in the cache or return false (network request)
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
