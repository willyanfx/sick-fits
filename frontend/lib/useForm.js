import { useState } from 'react';

export default function useForm(initial = {}) {
  const [input, setInput] = useState(initial);

  function handleChange(e) {
    let { value, name, type } = e.target;

    if (type === 'number') {
      value = parseInt(value);
    }

    if (type === 'file') {
      [value] = e.target.files;
    }

    setInput({
      ...input,
      [name]: value,
    });
  }

  function resetForm() {
    setInput(initial);
  }

  function clearForm() {
    const blankState = Object.entries(input).map(([key, value]) => [key, '']);
    setInput(Object.fromEntries(blankState));
  }

  return { input, handleChange, resetForm, clearForm };
}
