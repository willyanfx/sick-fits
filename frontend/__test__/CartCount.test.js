import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import wait from 'waait';
import CartCount from '../components/CartCount';
import { fakeItem } from '../lib/testUtils';

describe('<CartCount>', () => {
  it('should render', () => {
    const { container, debug } = render(<CartCount count={10} />);
    expect(container).toMatchSnapshot();
  });
  it('should update via props', async () => {
    const { container, rerender } = render(<CartCount count={10} />);
    expect(container.textContent).toBe('10');
    expect(container).toHaveTextContent('10'); // same as before
    rerender(<CartCount count={12} />);

    await waitFor(() => {
      expect(container.textContent).toBe('12');
    });
  });
});
