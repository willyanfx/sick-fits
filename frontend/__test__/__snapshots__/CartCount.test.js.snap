// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount> should render 1`] = `
<div>
  <span
    class="CartCount__AnimationStyles-sc-xxvp4g-1 fgBkbW"
  >
    <div>
      <div
        class="CartCount__Dot-sc-xxvp4g-0 fvA-DmT"
      >
        10
      </div>
    </div>
  </span>
</div>
`;
